texk/ps2pk/ps2pk.1
.TH PS2PK 1 "30 August 2022"
.SH NAME
ps2pk \- creates a TeX pkfont from a type1 PostScript font
.SH SYNOPSIS
ps2pk
[ -v ]
[ -e\fIenc\fP ]
[ -X\fIxres\fP ]
[ -E\fIexpansion\fP ]
[ -S\fIslant\fP  ]
[ -P\fIpointsize\fP ]
[ -R\fIbaseres\fP ]
[ -r\fIaspectratio\fP  ]
[ -Y\fIyres\fP ]
[ -a\fIAFMfile\fP ]
[ -m\fImodename\fP ]
type1 [pkfont]
.SH DESCRIPTION
This program renders a given type1 PostScript font at a given pointsize
(default 10.0 points) and resolution (default 300dpi) into a TeX pkfont. 
.PP
To generate the pkfont ps2pk needs a valid type1 font file (for
example Utopia-Regular, Utopia-Regula.pfb or Utopia-Regula.pfa) and
its corresponding AFM-file (Utopia-Regular.afm) with the font metrics.
The program accepts both the binary type1 format (Utopia-Regula.pfb)
and its ASCII equivalent (Utopia-Regular or Utopia-Regula.pfa).  To
locate the files, ps2pk uses the kpathsea library (see the info page
on kpathsea). Additionally, pfb and pfa files are searched for in the
directory specified in the environment variable T1INPUTS, and afm
files are searched in AFMFONTS, if they are set (the -v flag will
report which filenames are looked for).
.PP
The program will make a pkfont in which the character codes are derived
from their AFM (Adobe Font Metrics) defined values.  This can be
overruled by specifying an explicit encoding file via the -e option.
The encoding file \fIenc\fP
should contain a valid PostScript encoding containing PostScript names
like /hyphen.  Here is an incomplete example (see afm2tfm for complete
syntax):

.RS
.nf
% This is the EC encoding.
/ECEncoding [          % now 256 chars follow
% 0x00
  /grave /acute /circumflex /tilde /dieresis /hungarumlaut
  /ring /caron /breve /macron /dotaccent /cedilla
  /ogonek /quotesinglbase /guilsinglleft /guilsinglright

  (Rest of lines omitted.)

% 0xF0
  /eth /ntilde /ograve /oacute /ocircumflex /otilde
  /odieresis /oe /oslash /ugrave /uacute /ucircumflex
  /udieresis /yacute /thorn /germandbls ] def
.fi
.RE
.PP
Options and arguments:
.IP -a\fIAFMfile\fP 12
Overrules the name that is used to locate the AFM file.

.IP -e\fIenc\fP 12
Name of a file containing an encoding scheme 
(default \- the encoding named in the AFM-file is used). 
The actual PostScript name of the encoding scheme is written
as a "special" at the end of the pkfont.

.IP -E\fIextension\fP 12
The extension factor (real value, default 1.0).

.IP -m\fImodename\fP 12
A modename such as would be used by METAFONT (default "Unknown mode").

.IP -P\fIpointsize\fP 12
The desired pointsize (real value, default 10.0 points). 

.IP -R\fIbaseres\fP 12
The desired base resolution (integer value, default 300 dpi). 
If this differs from the value of \fIxres\fP, the appropriate
magnification will be shown in the "specials" at the end
of the pkfont. If possible, this will be expressed as a
magstep, otherwise as a real number. For example, a pkfont with an \fIxres\fP
of 329, using the default base resolution of 300,
will include the "special" text "mag=magstep(0.5)" at the
end of the font.

.IP -r\fIaspectratio\fP 12
The desired aspect ratio (expression, integerY/integerX, default 300/300). 
This will be shown in the "specials" at the end
of the pkfont.  Setting the aspect ratio creates a value for
\fIyres\fP but an explicit \fIyres\fP value will take precedence.
Argument must be in "" if spaces are left around `/'.

.IP -S\fIslant\fP 12
The slant (real value, default 0.0).

.IP -X\fIxres\fP 12
The resolution in the X direction (integer value, default 300 dpi). 

.IP -Y\fIyres\fP 12
The resolution in the Y direction (integer value, default the value
of \fIxres\fP). If \fIyres\fP differs from \fIxres\fP, a "special"
text giving the aspect ratio will be written at the end of the
pkfont.

.IP -v 12
Verbose flag. (Tells what the program is doing.)

.IP type1 12
The name of the PostScript type1 font.  The name of the AFM-file will be
constructed from this name by removing the extension (if supplied) and
adding ".afm".  The PostScript "FontName" is extracted from the
AFM-file and written into a "special" at the end of the pkfont.

.IP [pkfont] 12
The name of the resulting pkfont can be overruled with this name.  The
default name of pkfont is derived from the basename of the type1
font, the pointsize and \fIxres\fP.  For example
`ps2pk -P17.28 Utopia-Regular' will result in `Utopia-Regular17.300pk'.
An explicit value for the name of the pkfont is necessary when
the type1 font name already shows the point size, otherwise
the \fIpointsize\fP value is catenated on to the pkfont basename
which is usually not what is wanted.

.PP
The following PK "specials" provide an internal check on the
characteristics of the pkfont, in accordance with the recommendations
of the TeX Working Group on the TeX Directory Standard (TWG-TDS):
.br
"fontid=\fIFontName\fP", "codingscheme=\fIEncoding\fP", "fontfacebyte", "mag",
"mode=(ps2pk)\fImodename\fP", "pixels-per-inch", and "aspect-ratio" if it
is other than unity.

.SH ACKNOWLEDGEMENT
This program uses the Type1 hinting and rendering software IBM 
has donated to the X Consortium.

.SH SEE ALSO
.nf
`METAFONT: The Program', Donald Knuth.
`The GFtoPK processor', belonging to the METAFONTware.
`Adobe Font Metric Files', Version 3.0, PostScript Developer
Support Group.
afm2tfm(1)
pk2bm(1)
kpsewhich(1)
info kpathsea

.SH VERSION
1.4 (January 1994)

.SH AUTHOR
Piet Tutelaers,
modified for ps2pkm by Pierre A. Mackay.
This manpage was adapted to teTeX by Frank Kuester.
.PP
Maintained in TeX Live; please send bug reports or other correspondence
to tex-k@tug.org (https://lists.tug.org/tex-k).


ChangeLog
2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2020-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>


texk/ps2pk/ps2pk.c
/*
 * NAME
 *	ps2pk - creates a PK font from a type1 PostScript font
 * SYNOPSIS:
 *	ps2pk [options] type1 [pkname]
 * DESCRIPTION
 *	This program renders a given type1 PostScript font at a given
 *	pointsize (default 10.0 points) and resolution (default 300dpi)
 *      into a TeX PK font.
 *
 *      To generate the PK font ps2pk needs a valid type1 fontname (for
 *      example Utopia-Regular.pfa) and its corresponding AFMfile
 *	(Utopia-Regular.afm). The program accepts both the MSDOS binary
 *	type1 format (Utopia-Regula.pfb) and its UNIX ASCII equivalent
 *	(Utopia-Regula.pfa). The resulting PK font does contain all
 *	characters for which the C-code is non negative. This can be
 *	overruled by specifying an explicit <encoding> via the -e<encoding>
 *	option. Character codes in AFM files are specified as integer values
 *	after the C, for example Utopia-Regular.afm provides:
 *	   C 251 ; WX 601 ; N germandbls ; B 22 -12 573 742 ;
 *	   C  -1 ; WX 380 ; N onesuperior ; B 81 272 307 680 ;
 *
 *	The `germandbls' has code 251 and `onesuperior' -1 which means
 *	that this character is not part of the default AdobeStandard-
 *	Encoding scheme. To get this `onesuperior' you have to add it to
 *	an encoding vector as described in the PostScript manuals. But of
 *	course it is better to use standard encodings, like the one provided
 *	with the new psfonts. This `TeXBase1Encoding' encoding does contain
 *	all names of characters as defined in the 256 character TeX fonts.
 *
 *	The location of files for ps2pk, since version 1.5, has been
 *	improved using PostScript resource databases. These databases
 *	can be created using Adobe's mkpsres(1) utility. On UNIX the
 *	default name is PSres.upr and can be found in directories like
 *	/usr/lib/DPS depending on the machine you are using. The
 *	following example of an UPR file contains three resources
 *	   PS-Resources-1.0
 *	   Encoding
 *	   FontAFM
 *	   FontOutline
 *	   .
 *	   //usr/local/tex/fonts/type1faces
 *	   Encoding
 *	   ECEncoding=EC.enc
 *	   .
 *	   FontAFM
 *	   Courier=ResidentFonts/Courier.pfa
 *	   .
 *	   FontOutline
 *	   Courier=ResidentFonts/Courier.afm
 *	   .
 *
 *	When this UPR is in the PSRESOURCEPATH ps2pk can make an EC encoded
 *	PK font for Courier with:
 *
 *	   ps2pk -eECEncoding Courier
 *
 *	It is also possible without the UPR file but then you have to
 *	set T1INPUTS to the directories containing the files you need,
 *	for example:
 *
 *	   setenv T1INPUTS /usr/local/tex/fonts/type1faces//
 *
 *	and to run ps2pk by providing the proper filenames:
 *
 *	   ps2pk -eEC.enc Courier.pfa
 *
 *	In both cases the result will be the PK font Courier10.300pk.
 *	
 *	PostScript resource databases eliminate the need for recursive
 *	directory search, this is done by mkpsres(1) instead. And all
 *	resources can be stored in one directory per computer keeping 
 *	maintenance easy. Another advantage is the mapping of internal
 *	PostScript names, which are case sensitive and not restricted to
 *	MSDOS limitations, to external platform dependant filenames in
 *	a very natural way. On a MSDOS computer psres.dpr (the equivalent
 *	of PSres.upr) could be:
 *	   PS-Resources-1.0
 *	   Encoding
 *	   FontAFM
 *	   FontOutline
 *	   .
 *	   Encoding
 *	   TeXBase1Encoding=c:\emtex\dvips\8r.enc
 *	   .
 *	   FontAFM
 *	   Courier=c:\psfonts\com_____.pfb
 *	   .
 *	   FontOutline
 *	   Courier=c:\psfonts\afm\com_____.afm
 *	   .
 *
 *	Given this psres.dpr resource database
 *
 *	   ps2pk -eTeXBase1Encoding Courier
 *
 *	will create Courier10.300pk. Unfortunately MSDOS will abbreviate
 *	this name to COURIER1.300. Mkpsres(1) is an improved version
 *	of Adobe's makepsres(1) that runs on UNIX and MSDOS. This program
 *	will treat PFB fonts and PFM fonts correctly and recognize
 *	filenames ending with ".enc" as encoding vectors.
 *
 *	If the name of <type1font> does not contain an explicit extension
 *	or directory separator it is considered as the name of a resource.
 *	Otherwise as a filename. You can not mix up resource names and file-
 *	names. Filenames must have an extension!
 *
 * OPTIONS AND ARGUMENTS
 *       -a<AFM>        The name of the AFM resource or file can be
 *			overruled with this option.
 *       -e<encoding>   The encoding scheme (default the encoding from the 
 *                      AFM file is used).
 *	 -E<extension>	The extension factor (real value, default 1.0).
 *	 -O		Create old checksums (for compatibility)
 *	 -P<pointsize>	The desired pointsize (real value, default 10.0
 *			points). PK fonts created with a value different
 *			to 10.0 will cause dvips warnings `dvips: Design
 *			size mismatch in font ...'.
 *	 -R<baseres>	The base resolution (integer value, default 300
 *			dpi). Only used for creating METAFONT compatible 
 *			identification strings in PK postamble.
 *	 -S<slant>	The slant (real value, default 0.0).
 *	 -X<xres>	The resolution (integer value) in the X direction 
 *			(default 300 dpi).
 *	 -Y<yres>	The resolution (integer value) in the Y direction 
 *			(defaults to the value of <xres>).
 *	 -d		Debug stat() calls during recursive path searching
 *	 -v		Verbose flag. (Tells what the program is doing.)
 *	
 *	 type1font	The name of the PostScript type 1 font. When no
 *                      explicit AFMfile is given via the -a option the
 *                      name of the AFMfile will be constructed from this
 *                      name by removing the extension (if supplied) and
 *                      adding ".afm". 
 *	 [pkname]	The name of the resulting PK font can be overruled
 *			with this name. The default name of the PK font is
 *			derived from the basename of the type1 font, the
 *			pointsize and <xres>. For example:
 *			   ps2pk -P17.28 Utopia-Regular
 *			will result in:
 *				Utopia-Regular17.300pk
 *
 * ENVIRONMENTS
 *
 *   PSRESOURCEPATH
 *	The path used for locating PS resource database files (PSres.upr
 *	on UNIX and psres.dpr on MSDOS). The default of the path will be
 *	shown by running ps2pk with the -v flag and a non-existant resource
 *	name.
 *
 *   T1INPUTS
 *	The path used for locating PS type1 fonts, their AFM files and
 *	encoding vectors. Path may contain '//' (UNIX) or '!!' (MSDOS) to
 *	force ps2pk to look into subdirectories. Using PSRESOURCEPATH
 *	instead of T1INPUTS avoids recursion. The default of the path will
 *	be shown by running ps2pk with the -v flag and a non-existant
 *	filename.
 *	
 * ACKNOWLEDGEMENTS
 *	This program uses the type1 hinting and rendering software IBM 
 *	has donated to the X-consortium.
 *
 *      Since version 1.5 the lookup of PS resources as described in
 *      `Programming the Display PostScript System with X (Appendix A:
 *      Locating PostScript Resources)' from Adobe Systems is used.
 * SEE ALSO
 *	``METAFONT: The Program'', Donald Knuth.
 *	``The GFtoPK processor'', belonging to the METAFONTware.
 *	afm2tfm(1)
 *      mkpsres(1)
 *	mtpk(1)
 *	pk2bm(1)
 *	pkfonts(1)
 * AUTHOR
 *	Piet Tutelaers
 */

#include "basics.h"	/* basic definitions and fatal() */
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/* PostScript Resource lookup functions */
/* 
#include "PSres.h"	
*/
#include "filenames.h"	/* newname(), extension(), ps_resource() */
#include "psearch.h"	/* path search functions */
#include "pkout.h"	/* PK output functions */

/* begin interface to type 1 software */
#include "ffilest.h"

#include "objects.h"
#include "spaces.h"
#include "paths.h"
#include "regions.h"
#include "t1stdio.h"
#include "util.h"
#include "fontfcn.h"

FontScalableRec vals;
FontEntryRec entry;
#define Succesful	85
/* end interface to type 1 software */

char *encfile = NULL, *afmfile;

char ps2pk_args[MAXSTRLEN] = "none";	/* essential ps2pk arguments */

#define POINTSPERINCH 72.27
#define DEFAULTRES 300
#define DEFAULTPOINTSIZE 10.0

double pointsize = 0.0; /* wanted pointsize */
int W, H,	        /* width and height of character glyph */
    base_resolution = DEFAULTRES,
    x_resolution = 0, 
    y_resolution = 0;

int verbose = 0, debug = 0;

/* Provide old (-O flag) and new (default) checksum function */
static uint32_t checksum(encoding, int [256]);
static uint32_t old_checksum(encoding, int [256]);
uint32_t (*pchecksum)(encoding, int [256]) = checksum;

/* Prototypes */
static int next_pixel(void);
#if 0
static void print_pixmap(void);
#endif
static void first_pixel(CharInfoRec *);
static int32_t TFMwidth(int);
static int h_escapement(int);
static void add_option(const char *, const char *);

int main(int argc, char *argv[])
{
   char c;
   int done, i;
   const char *myname = "ps2pk";
   char *psname, *psbasename, *afmname = NULL,
	*encname = NULL, *psfile = NULL, *psfilebn, pkname[80],
	*AFM_fontname = NULL,
	*encodingscheme = NULL;

   FontPtr fontptr;
   unsigned char glyphcode[1]; /* must be an array */
   CharInfoRec *glyphs[1];
   unsigned int count;
   int charcode, rc = -1, charno;
   char comment[256];
   long cs;
   encoding ev;
   int WX[256];

   float efactor = 1.0, slant = 0.0;

#ifdef KPATHSEA
   kpse_set_program_name(argv[0], "ps2pk");
#endif

   while (--argc > 0 && (*++argv)[0] == '-') {
      done=0;
      while ((!done) && (c = *++argv[0]))  /* allow -bcK like options */
      	 switch (c) {
      	 case 'a':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    afmname = argv[0]; 
	    done = 1;
      	    break;
      	 case 'd':
      	    debug = 1;
	    break;
      	 case 'e':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    encname = argv[0]; 
	    add_option("-e", encname);
	    done = 1;
      	    break;
      	 case 'E':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    efactor = atof(argv[0]);
	    add_option("-E", argv[0]);
	    done = 1;
      	    break;
      	 case 'O':
      	    pchecksum = old_checksum;
	    add_option("-O", "");
	    break;
      	 case 'P':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    pointsize = atof(argv[0]); 
	    add_option("-P", argv[0]);
	    done = 1;
      	    break;
      	 case 'R':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    base_resolution = atoi(argv[0]); 
	    add_option("-R", argv[0]);
	    done = 1;
      	    break;
      	 case 'S':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    slant = atof(argv[0]);
	    add_option("-S", argv[0]);
	    done = 1;
      	    break;
      	 case 'X':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    x_resolution = atoi(argv[0]); done = 1;
	    if (y_resolution == 0) y_resolution = x_resolution;
	    add_option("-X", argv[0]);
      	    break;
      	 case 'Y':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
	    y_resolution = atoi(argv[0]); done = 1;
	    if (x_resolution == 0) x_resolution = y_resolution;
	    add_option("-Y", argv[0]);
      	    break;
      	 case 'v':
      	    verbose = 1;
	    break;
invalid:
      	 default:
      	    fatal("%s: %c invalid option\n", myname, c);
      	 }
      }

   if (argc < 1 || argc >2) {
      msg  ("ps2pk version " PACKAGE_VERSION " (1992-2016)\n");
      msg  ("Usage: %s [options] type1font [pkname]\n", myname);
      msg  ("options: -d -v -e<enc> -X<xres> -E<expansion> -S<slant>\n");
      fatal("options: -O -P<pointsize> -Y<yres> -a<AFM> -R<baseres>\n");
   }

   psname = argv[0]; argc--; argv++;

#ifndef KPATHSEA
   if (ps_resource(psname)) {
      /* if psname does neither contain a DIRSEP or `.' character
       * then find the filename of this resource in one of the PS
       * resource database(s) available in PSRESOURCEPATH.
       */
      char **names, **files; int i, count;

      psrespath = path(PSRESOURCEPATH, getenv("PSRESOURCEPATH"));
      if (verbose) msg("PSRESOURCEPATH := %s\n", psrespath);

      count = ListPSResourceFiles(psrespath, NULL,
   	"FontOutline", psname, &names, &files);
      if (count == 0)
	 fatal("No FontOutline resource found for %s\n", psname);
      if (verbose && count > 1) {
   	 msg("More than one FontOutline resource for %s\n", psname);
	 for (i = 0; i < count; i++) msg("\t<%s>\n", files[i]);
      }
      psfile = files[0];
      if (verbose) msg("Resource <%s>\n", psfile);
      psbasename = psname;
   }
   
   if (psfile != NULL) {
      char **names, **files; int count;

      if (afmname == NULL || ps_resource(afmname)) {
	 if (afmname == NULL) afmname = psname;
         count = ListPSResourceFiles(psrespath, NULL,
      	   "FontAFM", afmname, &names, &files);
         if (count == 0)
	    fatal("No FontAFM resource found for %s\n", afmname);
         if (verbose && count > 1) {
   	    msg("More than one FontAFM resource for %s\n", afmname);
	    for (i = 0; i < count; i++) msg("\t<%s>\n", files[i]);
   	 }
         afmfile = files[0];
         if (verbose) msg("Resource <%s>\n", afmfile);
      }
      else fatal("%s: invalid resource name\n", afmname);

      if (encname != NULL) {
	 if (!ps_resource(encname)) 
            fatal("%s: invalid Encoding resource\n", encname);
         count = ListPSResourceFiles(psrespath, NULL,
      	    "Encoding", encname, &names, &files);
         if (count == 0)
	    fatal("No Encoding resource found for %s\n", encname);
         if (verbose && count > 1) {
   	    msg("More than one Encoding resource for %s\n", encname);
	    for (i = 0; i < count; i++) msg("\t<%s>\n", files[i]);
         }
         encfile = files[0];
         if (verbose) msg("Resource <%s>\n", encfile);
      }
   }
   else { /* No PS resource database or no resource for type1font */
#endif   
#ifdef KPATHSEA
     psfile = kpse_find_file(psname,kpse_type1_format,0);
     if (!psfile) {
       /* kpse_find_file does not look for alternate suffixes */
       char *altname = newname(psname, ".pfb");
       psfile = kpse_find_file(altname, kpse_type1_format, 0);
       free(altname);
     }
     if (!psfile)
       fatal("%s: PS file %s not found!\n", kpse_invocation_name,
	     psname);
     if (afmname == NULL)
       afmname = newname(psname, ".afm");
     afmfile = kpse_find_file(afmname,kpse_afm_format,0);
     if (!afmfile)
       fatal("%s: AFM file %s not found!\n", kpse_invocation_name,
	     afmname);
     
     if (encname) {
       encfile = kpse_find_file(encname,kpse_enc_format,0);
       if (!encfile)
	 fatal("%s: enc file %s not found!\n", kpse_invocation_name,
	       encname);
     }
     else
       encfile = NULL;
     printf("psfile = %s\n", psfile);
     printf("afmfile = %s\n", afmfile);
     if (encfile) printf("encfile = %s\n", encfile);
#else
     t1inputs = path(T1INPUTS, getenv("T1INPUTS"));
     if (verbose) msg("T1INPUTS := %s\n", t1inputs);
     
     psfile = search_file(t1inputs, psname, FATAL);
     if (verbose) msg("Font file <%s>\n", psfile);
     
     if (afmname ==  NULL) afmname = newname(psfile, ".afm");
     afmfile = search_file(t1inputs, afmname, FATAL);
     if (verbose) msg("AFM file <%s>\n", afmfile);
     
     /* get encoding when -e<enc> provided */
     if (encname) {
       encfile = search_file(t1inputs, encname, FATAL);
       if (verbose) msg("Encoding file <%s>\n", encfile);
     }
#endif
     psbasename = basename(psname, extension(psname));
#ifndef KPATHSEA
   }
#endif
   psfilebn = basename(psfile, NULL);

   if (pointsize == 0.0) pointsize = DEFAULTPOINTSIZE;
   if (x_resolution == 0) x_resolution = DEFAULTRES;
   if (y_resolution == 0) y_resolution = x_resolution;
   if (verbose)
      msg("Loading encoding vector from %s ...",
	 encname? encname: basename(afmfile, NULL));
   getenc(&AFM_fontname, &encodingscheme, ev, WX);
   if (efactor != 1.0)
      for (i=0; i < 256; i++) {
         if (ev[i] == NULL) continue;
         WX[i] = WX[i] * efactor + 0.5;
      }
   if (verbose) msg(" done\n");
   
   if (argc == 1) strcpy(pkname,  argv[0]);
   else 
      sprintf(pkname, "%s%d.%dpk", psbasename,
         (int) (pointsize + 0.5), x_resolution);
    
   /* next values are needed! */
   vals.x =     x_resolution;
   vals.y =     y_resolution;
   vals.point = 10.0 * pointsize + 0.5; /* DECIPOINTS */
   vals.pixel = pointsize * y_resolution / POINTSPERINCH + 0.5;
   
   if (verbose) msg("Checking type1 font %s ...", psfilebn);
   Type1RegisterFontFileFunctions();
   if (verbose) msg(" done\n");

	/* next line prevents UNIX core dumps */
   entry.name.name = "-adobe-utopia-medium-r-normal--0-0-0-0-p-0-iso8859-1";
   if (verbose)
      msg("Creating character glyphs for %s ...", psfilebn);
   rc = Type1OpenScalable(ev, &fontptr, 0, &entry, 
            psfile, &vals, 0, 0, efactor, slant);
   if (rc != Succesful) 
      fatal("Type1OpenScalable error (result: %d) on %s \n",  rc, psfile);
   if (verbose) msg(" done\n");

   pk_open(pkname);
   sprintf(comment, "%s rendered at %f points", psbasename, pointsize);
   cs = pchecksum(ev, WX);
   pk_preamble(comment, pointsize, cs, x_resolution, y_resolution);
   charno = 0;
   if (verbose) msg("Creating %s from %s\n", pkname, psfilebn);
   for (charcode = 0; charcode < 256; charcode++) {
      if (ev[charcode] == NULL) continue;
      glyphcode[0] = charcode;
      (fontptr->get_glyphs)(fontptr, 1, glyphcode, 0, &count, glyphs);
      if (count > 0) {
         if (debug) {
            msg("'%03o ", charcode); charno++;
            if (charno == 8) { 
               msg("%c", '\n'); charno = 0;
            }
         }
	 first_pixel(glyphs[0]); /* assigns W and H */
         pk_char(charcode,			   /* charcode */
            TFMwidth(WX[charcode]), 		   /* TFMwidth */
            h_escapement(WX[charcode]),		   /* h_escapement */
            W, H,  				   /* width and height */
            - glyphs[0]->metrics.leftSideBearing,  /* h_offset */
            glyphs[0]->metrics.ascent,             /* v_offset */
            next_pixel);			   /* pixel generator */
      }
   }
   if (debug) msg("%c", '\n');
   ps2pk_postamble(psbasename, encodingscheme,
              base_resolution, x_resolution, y_resolution, pointsize,
	      ps2pk_args);
   pk_close();
   exit(0);
}

/*
 * The checksum should garantee that our PK file belongs to the correct TFM
 * file! Exactly the same as the afm2tfm (dvips5487) calculation.
 */
static uint32_t old_checksum(encoding ev, int width[256])
{
   int i;
   uint32_t s1 = 0, s2 = 0;
   char *p ;

   for (i=0; i<256; i++) {
      if (ev[i] == NULL) continue;
      s1 = (s1<<1) ^ width[i];                   /* left shift */
      for (p=ev[i]; *p; p++)
	 s2 = s2 * 3 + *p ;
   }
   return (s1<<1) ^ s2 ;
}

/*
 * The new checksum algorithm.
 */
static uint32_t checksum(encoding ev, int width[256])
{
   int i;
   uint32_t s1 = 0, s2 = 0;
   char *p ;

   for (i=0; i<256; i++) {
      if (ev[i] == NULL) continue;
      s1 = ((s1<<1) ^ (s1>>31)) ^ width[i];   /* cyclic left shift */
      for (p=ev[i]; *p; p++)
	 s2 = s2 * 3 + *p ;
   }
   return (s1<<1) ^ s2 ;
}

static int row, col;
static int data, bitno;
static unsigned char *p_data;

static int
next_pixel(void)
{
   int pixel;

   while (row < H) {
       if (col++ < W) { 
           if (bitno == 0) { data = *p_data++; bitno = 8; }
           pixel = data & 0x80? BLACK: WHITE; 
           data <<= 1; bitno--; return pixel;
       }
       col = 0; row++; bitno = 0;
   }
   fatal("Not that many pixels!\n");
} 

static void
first_pixel(CharInfoRec *G)
{
   row = col = 0;
   p_data = (unsigned char *) G->bits;
   W = G->metrics.rightSideBearing - G->metrics.leftSideBearing;
   H = G->metrics.ascent + G->metrics.descent;
   bitno = 0; 
}

#if 0
static void
print_pixmap(void)
{
   int c, r;
   unsigned char *save_p_data;

   save_p_data = p_data;
   if (W*H == 0) return; /* easy */
   for (r = 0; r < H; r++) {
      for (c = 0; c < W; c++)
	 if (next_pixel() == BLACK) msg("%c", 'X');
	 else msg("%c", '.');
      msg("%c", '\n');
   }
   /* reset data for scanning pixmap */
   p_data = save_p_data;
   bitno = 0; row = 0; col = 0;
}
#endif

/* Next function computes the width as a fix_word. A fix_word is 
   an integer representation for fractions. The decimal point is 
   left from bit 20. (The method is `stolen' from afm2tfm.) */
 
static int32_t
TFMwidth(int wx)
{  
   return (((wx  / 1000) << 20) +
           (((wx % 1000) << 20) + 500) / 1000) ;
}

static float HXU = -1.0; /* horizontal pixels per design unit */

/* the horizontal escapent is the number of pixels to next origin */
static int
h_escapement(int wx)
{
   if (HXU == -1.0) 
      HXU = (pointsize * x_resolution) / 72270.0;
   return wx * HXU +  0.5;
}

/* add ps2pk option to string */
static void
add_option(const char *option, const char *value)
{
   static char *p_args = NULL;

   if (p_args == NULL) {
      p_args = ps2pk_args;
      sprintf(p_args, "%s%s", option, value);
   }
   else
      sprintf(p_args, " %s%s", option, value);
   p_args+= strlen(p_args);
}

/* Next stuff is needed by type1 rendering functions */

int CheckFSFormat(int format, int fmask, int *bit, int *Byte, int *scan,
                  int *glyph, int *image)
{
       *bit = *Byte = 1;
       *glyph = *scan = *image = 1;
       return Successful;
 
}
 
intptr_t MakeAtom(const char *p, unsigned int len, int foo)
{
       return (intptr_t)p;
}

#if 0
void GetClientResolutions(int *resP)
{
       *resP = 0;
}
#endif

void *Xalloc(size_t size)
{
   void *p;
   p = malloc(size);
   if (p == NULL) fatal("Out of memory\n");
   return p;
}
 
void Xfree(void *p)
{
       free(p);
}
